src/components/smart/Table.vue

```diff
@@ -1,69 +1,321 @@
+    list: Item[]
+    /** The headings of the table */
+    headings?: CellHeading[]
+    /** Whether to show the search bar */
+    searchBar?: {
+      /** Whether to debounce the search query event */
+      debounce?: number
+      placeholder?: string
+    }
+    /** Whether to show the checkbox column
+     * This will be overriden if custom implementation for body slot is provided
+     */
+    checkbox?: boolean
+
+    selectedRows?: Item[]
+    /** Whether to enable sorting */
+    sort?: {
+      /** The key to sort the list by */
+      key: string
+      direction: Direction
+    }
+
+    /** Whether to enable pagination */
+    pagination?: {
+      totalPages: number
+    }
+
+    /** Whether to show the spinner */
+    spinner?: {
+      enabled: boolean
+      duration?: number
+    }
+  }>(),
+  {
+    showYBorder: false,
+    search: undefined,
+    checkbox: false,
+    sort: undefined,
+    selectedRows: undefined,
+  },
+)
 
 const emit = defineEmits<{
   (event: "onRowClicked", item: Item): void
+  (event: "update:list", list: Item[]): void
+  (event: "search", query: string): void
+  (event: "update:selectedRows", selectedRows: Item[]): void
+  (event: "pageNumber", page: number): void
 }>()
 
+// Pagination functionality
+const page = ref(1)
+
+enum PageDirection {
+  Previous,
+  Next,
+}
+
+const changePage = (direction: PageDirection) => {
+  const isPrevious = direction === PageDirection.Previous
+  if (
+    (isPrevious && page.value > 1) ||
+    (!isPrevious && page.value < props.pagination!.totalPages)
+  ) {
+    page.value += isPrevious ? -1 : 1
+  }
+
+  emit("pageNumber", page.value)
+}
+
+// The working version of the list that is used to perform operations upon
+const workingList = useVModel(props, "list", emit)
+
+// Spinner functionality
+const isSpinnerEnabled = ref(false)
+const showSpinner = (duration: number = 500) => {
+  isSpinnerEnabled.value = true
+  setTimeout(() => {
+    isSpinnerEnabled.value = false
+  }, duration)
+}
+
+watch(
+  () => props.spinner,
+  () => {
+    if (props.spinner?.enabled === true) {
+      showSpinner(props.spinner.duration)
+    }
+  },
+)
+
+// Checkbox functionality
+const selectedRows = useVModel(props, "selectedRows", emit)
+
+watch(workingList.value, (updatedList) => {
+  if (props.checkbox) {
+    updatedList = updatedList.map((item) => ({
+      ...item,
+      selected: false,
+    }))
+  }
+})
+
 const onRowClicked = (item: Item) => emit("onRowClicked", item)
+
+const isRowSelected = (item: Item) => {
+  const { selected, ...data } = item
+  return selectedRows.value?.some((row) => isEqual(row, data))
+}
+
+const toggleRow = (item: Item) => {
+  item.selected = !item.selected
+  const { selected, ...data } = item
+
+  const index = selectedRows.value?.findIndex((row) => isEqual(row, data)) ?? -1
+
+  if (item.selected && !isRowSelected(data)) selectedRows.value!.push(data)
+  else if (index !== -1) selectedRows.value?.splice(index, 1)
```

Comment:
It might be helpful to represent these as block statements (brackets) for readability.